### PR TITLE
Support parallel chains in HMC

### PR DESCRIPTION
**Pull request by neerajprad**
Addresses #1093. 

This uses the python `multiprocessing` library to spin up separate python interpreters each running its own MCMC chain in parallel. I had to make some modifications to the `logging` module to make it work with parallel runners by using a logging handler that uses a `multiprocess.Queue` for logging events. On windows (also when we want to use CUDA), only the "spawn" context is available (no "fork"), which puts certain additional constraints on both the internals as well as the model callable, in that a lot of the code that is exercised needs to be picklable.

Some notes:
 - There is no change to the external interface. Multiprocessing with parallel workers will only kick in when `num_chains > 1`. We should be able to swap out the internals later to use tensor based parallelization, for instance.
 - The majority of change is restricted to the external `MCMC` wrapper. The only internal change is that the poutine wrapping of the model (`poutine.broadcast` or `poutine.enum`) happens in the setup method instead of the constructor. The reason for this is that our poutine code uses a lot of closures which makes it difficult to pickle poutine wrapped models. By moving this to `setup`, the poutine wrapping happens inside each of the processes so as not to be an issue. Once we make our handlers picklable, poutine wrapped models can be sent directly by the user.
 - Logging is handled asynchronously using a separate thread that consumes events from the logging `queue` into which each of the workers put in diagnostics / other logging messages. I found this to be the only reliable way of getting sane logging output with processes running in parallel. Most of the changes in `logger.py` relate to this change.

e.g. invocation (8 schools example)
```bash
   [pytorch-36] npradhan@npradhan:~/workspace/pyro_dev/pyro (hmc-chain-tqdm)
  $ python examples/eight_schools/mcmc.py --num-chains 4
Warmup [0]:  24%|██████▊                     | 305/1250 [00:07<00:23, 41.01it/s, Step size=0.435, Acceptance rate=0.941]
Warmup [1]:  24%|██████▌                     | 294/1250 [00:07<00:22, 42.32it/s, Step size=0.401, Acceptance rate=0.946]
Warmup [2]:  27%|███████▊                     | 337/1250 [00:07<00:17, 51.50it/s, Step size=0.59, Acceptance rate=0.947]
Warmup [3]:  25%|███████                     | 318/1250 [00:07<00:19, 47.73it/s, Step size=0.402, Acceptance rate=0.931]
```

#### TODOS:
 - Check execution on:
   - [x] Windows, Python 2/3 
   - [x] Mac OS X, Python 2/3
   - [x] Linux, Python 2/3
 - [ ] Check if CUDA backend works with `spawn` context - will need to redesigned due to https://github.com/pytorch/pytorch/issues/10375
 - [x] Deterministic (different for each chain) random number seeding w.r.t. single chain.
 - [x] Clean exit of child processes with KeyboardInterrupt.
 - [x] Close all processes and cleanup if there is an exception in one of the child processes.
 - [x] Pickling issue on dynamically generated torch distributions - #1456 

#### Later PR:
 - [ ] Chain level stats / diagnostics.

### CODE REVIEW

**Comment by fritzo** [comment]
@eb8680 it looks like `functools.partial(my_poutine, **kwargs)` is pickleable, whereas the closure `lambda f: my_poutine(f, **kwargs)` is not pickleable. Can we simply switch from closures to `functools.partial` in our poutines?

**Comment by neerajprad** [comment]
As I discussed with @eb8680, I have worked around the poutine pickling issue, so I am not currently blocked on that. It will be great to have `Messenger` and handlers like `condition`, `enum` etc. be picklable eventually, so that poutine wrapped models can be sent as args directly.

**Comment by neerajprad** [comment]
I am looking at some cleanup related issue in linux, but this should be good to review now. I will make changes for CUDA in a separate PR.

**Comment by neerajprad** [comment]
@eb8680 - Could you help test the baseball example on windows (just running `python examples/baseball.py`)? Multiprocessing is quite different in windows, so just wanted to make sure that we know of any issues beforehand.

File: pyro/infer/mcmc/logger.py [hunk 1]
```diff
@@ -93,25 +172,35 @@ def initialize_progbar(warmup_steps, num_samples, min_width=100, max_width=120,
                         position=pos, file=sys.stderr, disable=disable)
 
     if getattr(progress_bar, "ncols", None) is not None:
-        progress_bar.ncols = min(min_width, progress_bar.ncols)
-        progress_bar.ncols = max(max_width, progress_bar.ncols)
+        progress_bar.ncols = max(min_width, progress_bar.ncols)
+        progress_bar.ncols = min(max_width, progress_bar.ncols)
```
Comment (neerajprad) [inline pyro/infer/mcmc/logger.py]: This was just wrong earlier, so just correcting that.

**Comment by eb8680** [comment]
> Could you help test the baseball example on windows

Seems to work correctly for me, at least with Python 3.

**Comment by neerajprad** [comment]
> I'd recommend adding a disclaimer somewhere about a relative lack of windows testing.

Thanks, @eb8680. I will add that disclaimer. KeyboardInterrupt should result in exiting from the main thread very fast, and we shouldn't be leaving orphan processes behind (it might take a few seconds for them to be cleaned up though). I have tested that the cleanup happens as expected in OS X, and will run a few tests on linux too.

**Comment by neerajprad** [comment]
Thanks for reviewing! This should be good to merge.

**Pull request merged**